Source/WebCore/page/LocalFrame.cpp

```diff
@@ -242,6 +245,16 @@ bool LocalFrame::isRootFrame() const
     return m_isRootFrame;
 }
 
+LocalFrame& LocalFrame::rootFrame()
+{
+    if (auto* localMainFrame = dynamicDowncast<LocalFrame>(mainFrame()))
+        return *localMainFrame;
+    auto* iterator = this;
+    while (auto* localParent = dynamicDowncast<LocalFrame>(iterator->tree().parent()))
```

Comment:
Can't we have `Page` or someone hold onto the root frame, rather than having to climb the tree every time?

Reply:
The Page does have m_rootFrames, but we don't know which root frame is this frame's root frame.  For Frame::mainFrame we store a pointer m_mainFrame on each Frame.  We could do the same for this if it were common enough to be worth the memory use.

Reply:
This is fixed after https://github.com/WebKit/WebKit/pull/27823.